include/nbl/video/IGPURayTracingPipeline.h

```diff
@@ -0,0 +1,100 @@
+                CAPTURE_INTERNAL_REPRESENTATIONS = base_flag(CAPTURE_INTERNAL_REPRESENTATIONS),
+                FAIL_ON_PIPELINE_COMPILE_REQUIRED = base_flag(FAIL_ON_PIPELINE_COMPILE_REQUIRED),
+                EARLY_RETURN_ON_FAILURE = base_flag(EARLY_RETURN_ON_FAILURE),
+                LINK_TIME_OPTIMIZATION = base_flag(LINK_TIME_OPTIMIZATION),
+                RETAIN_LINK_TIME_OPTIMIZATION_INFO_BIT_EXT = base_flag(RETAIN_LINK_TIME_OPTIMIZATION_INFO),
+                RAY_TRACING_SKIP_TRIANGLES_BIT_KHR = 1<<12,
+                RAY_TRACING_SKIP_AABBS_BIT_KHR = 1<<13,
+                RAY_TRACING_NO_NULL_ANY_HIT_SHADERS_BIT_KHR = 1<<14,
+                RAY_TRACING_NO_NULL_CLOSEST_HIT_SHADERS_BIT_KHR = 1<<15,
+                RAY_TRACING_NO_NULL_MISS_SHADERS_BIT_KHR = 1<<16,
+                RAY_TRACING_NO_NULL_INTERSECTION_SHADERS_BIT_KHR = 1<<17,
+                RAY_TRACING_SHADER_GROUP_HANDLE_CAPTURE_REPLAY_BIT_KHR = 1<<19,
+                RAY_TRACING_ALLOW_MOTION_BIT_NV = 1<<20,
+            };
+            #undef base_flag
+
+            inline SSpecializationValidationResult valid() const
+            {
+                if (!layout)
+                  return {};
+
+                SSpecializationValidationResult retval = {
+                    .count=0,
+                    .dataSize=0,
+                };
+                const bool valid = pipeline_t::SCreationParams::impl_valid([&retval](const IGPUShader::SSpecInfo& info)->bool
+                {
+                    const auto dataSize = info.valid();
+                    if (dataSize<0)
+                        return false;
+                    else if (dataSize==0)
+                        return true;
+
+                    const size_t count = info.entries ? info.entries->size():0x80000000ull;
+                    if (count>0x7fffffff)
+                        return {};
+                    retval += {.count=dataSize ? static_cast<uint32_t>(count):0,.dataSize=static_cast<uint32_t>(dataSize)};
+                    return retval;
+                });
+                if (!valid)
+                    return {};
+                return retval;
+            }
+
+            inline std::span<const IGPUShader::SSpecInfo> getShaders() const { return shaders; }
+
+            // TODO: Could guess the required flags from SPIR-V introspection of declared caps
+            core::bitflag<FLAGS> flags = FLAGS::NONE;
+        };
+
+        inline core::bitflag<SCreationParams::FLAGS> getCreationFlags() const { return m_flags; }
+
+        // Vulkan: const VkPipeline*
+        virtual const void* getNativeHandle() const = 0;
+
+        virtual std::span<uint8_t> getRaygenGroupShaderHandle() const = 0;
+        virtual std::span<uint8_t> getHitGroupShaderHandle(uint32_t index) const = 0;
+        virtual std::span<uint8_t> getMissGroupShaderHandle(uint32_t index) const = 0;
+        virtual std::span<uint8_t> getCallableGroupShaderHandle(uint32_t index) const = 0;
```

Comment:
problem is that spans are pointers and as such are not really copyable, the only way this would work stably is if you queried all handles up front, stored them in a member and then handed our pointers to members.

Also a span is 16 bytes (two pointers) whereas a reference is 8 bytes.

All of this would be fine, except ofc one wants/needs type-safety

what I would suggest is define a 
```cpp
struct SShaderGroupHandle final
{
   protected:
      uint8_t data[SPhysicalDeviceLimits::ShaderGroupHandleSize];
};
static_assert(sizeof(ShaderGroupHandle)==SPhysicalDeviceLimits::ShaderGroupHandleSize);
```
and then
```cpp
virtual SShaderGroupHandle& getRaygen() const 0;
... and so on, but wihtout GroupShaderHandle in the name to avoid tautologies ...
```